packages/beacon-node/src/chain/validation/attestation.ts

```diff
@@ -568,41 +572,116 @@ export function verifyHeadBlockAndTargetRoot(
 }
 
 /**
- * Get a state for attestation verification.
- * Use head state if:
- *   - attestation slot is in the same fork as head block
- *   - head state includes committees of target epoch
+ * Get a shuffling for attestation verification from the ShufflingCache.
+ * - if blockEpoch is attEpoch, use current shuffling of head state
+ * - if blockEpoch is attEpoch - 1, use next shuffling of head state
+ * - if blockEpoch is less than attEpoch - 1, dial head state to attEpoch - 1, and add to ShufflingCache
+ *
+ * This implementation does not require to dial head state to attSlot at fork boundary because we always get domain of attSlot
+ * in consumer context.
  *
- * Otherwise, regenerate state from head state dialing to target epoch
+ * This is similar to the old getStateForAttestationVerification
+ * see https://github.com/ChainSafe/lodestar/blob/v1.11.3/packages/beacon-node/src/chain/validation/attestation.ts#L566
  */
-export async function getStateForAttestationVerification(
+export async function getShufflingForAttestationVerification(
```

Comment:
What happens if this function is called twice at the same time for a checkpoint that requires regen? What happens if we ever raise the concurrency of regen queue to more than 1?

I think there should be an added layer of caching in the shuffling cache tracking active regen requests as pending promises. When multiple callers want the same pending shuffle, return the promise to all to always _guarantee_ that there will never be repeated regen for attestation verification.

Also this code is extremely sensitive and can blow up Lodestar easily. Placing this code in the /validation folder as a random function hides how critical it is. This code should live closer to regen and have the locks explained in the paragraph above backed into it, and comments explaining their cost.

Reply:
@dapplion thanks this is a great point :+1, I addressed in #6030